feat: Payment Terms Status report

ruthra-kumar opened the pull request:
## Payment Terms Status Report
This report aims to provide current status of payment terms based on the associated invoices against the SO.
All invoices associated with Sales Order are fetched and assigned to terms in FIFO method. Statuses shown in the report are calculated at runtime and does not affect the state in the database.
<img width="1552" alt="payment terms status" src="https://user-images.githubusercontent.com/3272205/148046071-e75363c6-8541-4144-a15e-ec63268dc7b2.png">

When Individual SO's are selected, a chart is used to show the payment and paid amount for each terms.
<img width="1552" alt="charts for Payment terms" src="https://user-images.githubusercontent.com/3272205/148046791-8a3d6e39-ebbc-498b-ac1a-a9f1f7b1f567.png">

Documentation can be found [**here**]( https://docs.erpnext.com/docs/v13/user/manual/en/accounts/payment_terms_status_report )

ruthra-kumar committed:
feat: Payment Terms Status report

 - calculate status at runtime for payment terms based on invoices
 - invoices are used in FIFO method

ruthra-kumar committed:
test: fix failing test case payment terms status

ruthra-kumar committed:
refactor: remove unused imports

ruthra-kumar committed:
test: qty and rate changed to remove need for fractional Nos

ruthra-kumar commented:
commenting to remove inactive flag

deepeshgarg007 commented on erpnext/selling/report/payment_terms_status_for_sales_order/payment_terms_status_for_sales_order.js:
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
```
```suggestion
// Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and contributors
```

deepeshgarg007 commented on erpnext/selling/report/payment_terms_status_for_sales_order/payment_terms_status_for_sales_order.py:
```diff
@@ -0,0 +1,211 @@
+			"read_only": 1,
+		},
+		{
+			"label": _("Submitted"),
```
```suggestion
			"label": _("Posting Date"),
```

deepeshgarg007 commented on erpnext/selling/report/payment_terms_status_for_sales_order/payment_terms_status_for_sales_order.py:
```diff
@@ -0,0 +1,211 @@
+			"read_only": 1
+		},
+		{
+			"label": _("Payment Term"),
```
```suggestion
			"label": _("Payment Terms Template"),
```

ruthra-kumar replied:
This field denotes the individual term in the template. I think, labelling it 'Payment Terms Template' would cause confusion.

deepeshgarg007 commented on erpnext/selling/report/payment_terms_status_for_sales_order/payment_terms_status_for_sales_order.py:
```diff
@@ -0,0 +1,211 @@
+		{
+			"label": _("Paid Amount"),
+			"fieldname": "paid_amount",
+			"fieldtype": "Currency",
```
Different companies could have different default currencies. Currently for all companies current from Global Defaults will be picked. You'll have to add a hidden column for current and use that as Options for all the Currency columns to avoid this

ruthra-kumar replied:
added new hidden column as suggested

deepeshgarg007 commented on erpnext/selling/report/payment_terms_status_for_sales_order/payment_terms_status_for_sales_order.py:
```diff
@@ -0,0 +1,211 @@
+			"fieldname": "name",
+			"fieldtype": "Link",
+			"options": "Sales Order",
+			"read_only": 1,
```
Is read-only applicable here? I mean here columns are by default read-only or some new feature has been introduced recently

ruthra-kumar replied:
removed.

deepeshgarg007 committed:
fix: Copyright info

ruthra-kumar committed:
refactor: currency field and code cleanup

ruthra-kumar committed:
fix: default to company currency in report output

ruthra-kumar committed:
refactor: create invoices list without if else

ruthra-kumar committed:
test: refactor and fix failing test case

ruthra-kumar committed:
test: added test for alternate currency

- Sales Order and Invoice will be submitted in USD with exchange rate
of 70 with the default company currency
- Report will display in defauly company currency

deepeshgarg007 committed:
Merge branch 'develop' into payment_terms_report

deepeshgarg007 commented:
@Mergifyio backport version-13-hotfix

letsfkccp commented:
How is this feature going? Thank you.
-Generate Invoice based on Payment Terms
